collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java

```diff
@@ -0,0 +1,352 @@
+                        Timber.i("Extremely long post is being split into multiple posts");
+                        multipartBuilder.addPart(MultipartBody.Part.createFormData("*isIncomplete*", "yes"));
+                        ++fileIndex; // advance over the last attachment added...
+                        break;
+                    }
+                }
+            }
+
+            MultipartBody multipartBody = multipartBuilder.build();
+
+            try {
+                OkHttpClient client = createOkHttpClient(credentials);
+                Request request = postRequest(uri, multipartBody);
+                Response response = client.newCall(request).execute();
+
+                messageParser = new ResponseMessageParser(
+                        response.toString(),
+                        response.code(),
+                        response.message());
+
+                discardEntityBytes(response);
+
+                if (response.code() != HttpURLConnection.HTTP_CREATED && response.code() != HttpURLConnection.HTTP_ACCEPTED) {
+                    return messageParser;
+                }
+            } catch (IOException e) {
+                Timber.e(e);
+                String msg = e.getMessage();
+                if (msg == null) {
+                    msg = e.toString();
+                }
+
+                throw new IOException(msg);
+            }
+        }
+
+        return messageParser;
+    }
+
+    private OkHttpClient createOkHttpClient(@Nullable HttpCredentialsInterface credentials) {
+        OkHttpClient.Builder builder = new OkHttpClient.Builder();
+
+        addCredentials(builder, credentials);
+
+        return builder
+                .connectTimeout(CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+                .writeTimeout(WRITE_CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+                .readTimeout(READ_CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+                .followRedirects(true)
+                .build();
+    }
+
+    private void addCredentials(OkHttpClient.Builder builder, @Nullable HttpCredentialsInterface credentials) {
+        if (credentials != null) {
+            final Map<String, CachingAuthenticator> authCache = new ConcurrentHashMap<>();
+            Credentials cred = new Credentials(credentials.getUsername(), credentials.getPassword());
+
+            BasicAuthenticator basicAuthenticator = new BasicAuthenticator(cred);
+            DigestAuthenticator digestAuthenticator = new DigestAuthenticator(cred);
+
+            DispatchingAuthenticator authenticator = new DispatchingAuthenticator.Builder()
+                    .with("digest", digestAuthenticator)
+                    .with("basic", basicAuthenticator)
+                    .build();
+
+            builder.authenticator(new CachingAuthenticatorDecorator(authenticator, authCache))
+                    .addInterceptor(new AuthenticationCacheInterceptor(authCache));
```

Comment:
I believe this takes care of pre-emptive auth and also handles the Ona case. What happens when there are two authenticators specified? Does it try the first and then the second if it's not supported? Does `DispatchingAuthenticator` take care of using the auth type specified in the auth challenge? It would be good to have a comment on the method with some of those details.

As @jd-alexander points out, there should also be a filter to only allow basic auth on ports 443, 8443. Throwing an exception which will eventually be shown to the user would be fine in the case of basic auth on HTTP.

Reply:
Had a quick private chat with @jd-alexander about testing and we thought the logic here would be worth adding tests for which I don't think exist yet. That is:
- If a server responds to a request with a 401 and requests digest auth, digest auth is used and the provided realm and nonce are used for further requests
- If an https server responds to a request with a 401 and requests basic auth, basic auth is used for further requests
- If an http server responds to a request with a 401 and requests basic auth without an https redirect, the client should not send auth headers
- The first `HEAD` request should not send auth headers. Once the client receives a 401 and finds out what auth is required by the server, then auth headers should be sent

I suspect that the interceptor as written will always send basic auth headers on the first `HEAD` request which shouldn't be the case. A test would help verify that either way.

Reply:
Hi @lognaturel,

As discussed, rather than writing tests I'm using proxy to intercept what happens so that I can answer the questions above.

_"If a server responds to a request with a 401 and requests digest auth, digest auth is used and the provided realm and nonce are used for further requests"_

a) Confirmed that the provided realm and nonce are used for further requests